File: R/langevitour.R
#' Langevin Tour
#'
#' Make a Langevin Tour HTML widget, which can be used to explore high-dimensional numerical datasets.
#'
#' The only required argument is \code{X}, the high-dimensional collection of points. The \code{group} argument is also commonly used so that groups of points can be distinguished by color. Further arguments adjust the appearance or provide advanced features.
#'
#' langevitour will by default not scale variables individually. If you want variables to be individually scaled, use something like \code{scale=apply(X,2,sd)*4}. Using the \code{scale} argument rather than modifying \code{X} directly ensures the plot axes within the widgets retain the original units.
#'
#' In Javascript, the langevitour object can be obtained using \code{document.getElementById(elementId).langevitour}. For example you could have a button that sets the state of a widget using \code{document.getElementById(elementId).langevitour.setState(desiredState)}.
#'
#' @param X The data to plot. A matrix of numeric data, or something that can be cast to a matrix. Rows will be shown as points in the widget. Columns are the variables of your data.
#'
#' @param group A group for each row in X, will be used to color points. A factor, or something that can be cast to a factor.
#'
#' @param name A name for each row in X.
#'
#' @param center Center for each variable. If omitted, the column means will be used.
#'
#' @param scale Scale for each variable. Scale +/- center will be the range of guaranteed visible data. If omitted, a reasonable default will be chosen, equal for all variables. (The default is the largest singular value of the centered X times 2.5.)
#'
#' @param extraAxes A matrix with each column defining a projection of interest. The columns of \code{X \%*\% extraAxes} will be presented as extra "variables".
#'
#' @param lineFrom A vector of row numbers. Draw lines starting at these rows.
#'
#' @param lineTo A vector of row numbers. Draw lines ending at these rows.
#'
#' @param lineColors Character vector. A CSS color for each line.
#'
#' @param axisColors Character vector. CSS colors for each variable and then each extra axis.
#'
#' @param levelColors Character vector. CSS colors for each level of \code{group}.
#'
#' @param colorVariation Number between 0 and 1. Individual points are given slightly different brightnesses. How strong should this effect be?
#'
#' @param pointSize Point radius in pixels. A single number, or a number for each row in X.
#'
#' @param subsample For speed, randomly subsample down to this many rows.
#'
#' @param state A JSON string, or an object that htmlwidgets will convert to the correct JSON. Initial widget state settings. The state of a widget can be obtained from its "further controls and information" pane. I am not going to guarantee that states will be compatible between versions of langevitour. Hint: Since JSON uses double quotes, surround the string in single quotes.
#'
#' @param width Width of widget in CSS units, for example "700px" or "100%".
#'
#' @param height Height of widget in CSS units, for example "600px" or "75vh".
#'
#' @param elementId An element ID for the widget, see \code{htmlwidgets::createWidget}.
#'
#' @param link A SharedData object from the crosstalk package to share selections and filters with other htmlwidgets. The data in this object is not used, just the keys and group name. The rows of \code{link$origData()} should correspond to the rows of X.
#'
#' @param link_filter TRUE or FALSE. If using crosstalk, should hiding groups in langevitour also cause them to be filtered in linked widgets?
#'
#' @return An htmlwidget object.
#'
#' @examples
#' library(palmerpenguins)
#' 
#' completePenguins <- na.omit(penguins[,c(1,3,4,5,6)])
#' scale <- apply(completePenguins[,-1], 2, sd)*4
#' 
#' langevitour(
#'     completePenguins[,-1], 
#'     completePenguins$species, 
#'     scale=scale, pointSize=2)
#'
#'
#' # An example setting the widget's initial state
#'  
#' langevitour(
#'     completePenguins[,-1], 
#'     completePenguins$species, 
#'     scale=scale, pointSize=2,
#'     state='{"guideType":"pca","labelInactive":["bill_length_mm"]}')
#'
#' @export
langevitour <- function(
        X, group=NULL, name=NULL, center=NULL, scale=NULL, 
        extraAxes=NULL, lineFrom=NULL, lineTo=NULL, lineColors=NULL,
        axisColors=NULL, levelColors=NULL, colorVariation=0.1, pointSize=1, subsample=NULL, 
        state=NULL, width=NULL, height=NULL, elementId=NULL,
        link=NULL, link_filter=TRUE) {
    
    # Ensure data is matrix
    
    X <- as.matrix(X)
    
    columnNames <- colnames(X)
    colnames(X) <- NULL
    rownames(X) <- NULL
    
    if (is.null(columnNames))
        columnNames <- paste0("V", seq_len(ncol(X)))
    
    
    # Grouping
    
    if (is.null(group))
        group <- rep("", nrow(X))
    
    names(group) <- NULL
    
    group <- as.factor(group)
    
    
    # Check for crosstalk to allow linked selections
    crosstalkGroup <- NULL
    crosstalkKey <- NULL
    dependencies <- NULL
    
    if (!is.null(link)) {
        assertthat::assert_that(crosstalk::is.SharedData(link))
        dependencies <- crosstalk::crosstalkLibs()
        crosstalkGroup <- link$groupName()
        crosstalkKey <- link$key()
    }
    
    assertthat::assert_that( assertthat::is.flag(link_filter) )
    
    
    # Check for problems (not exhaustive!)
    
    assertthat::assert_that( 
        ncol(X) >= 2,
        length(columnNames) == ncol(X),
        length(group) == nrow(X),
        length(pointSize) == 1 || length(pointSize) == nrow(X),
        is.null(name) || length(name) == nrow(X),
        length(lineFrom) == length(lineTo),
        is.null(lineColors) || length(lineColors) == length(lineFrom),
        all(lineFrom >= 1), 
        all(lineTo >= 1), 
        all(lineFrom <= nrow(X)), 
        all(lineTo <= nrow(X)),
        assertthat::noNA(X),
        assertthat::noNA(group),
        assertthat::noNA(name),
        assertthat::noNA(lineFrom),
        assertthat::noNA(lineTo))
    
    
    # Centering
    
    if (is.null(center))
        center <- colMeans(X, na.rm=TRUE)
    if (length(center) == 1)
        center <- rep(ncol(X), center)
    
    names(center) <- NULL
    
    assertthat::assert_that( length(center) == ncol(X) )
    
    
    # Scaling
    
    # Potentially expensive
    if (is.null(scale)) {
        X_centered <- sweep(X,2,center,"-")
        scale <- max(svd(X_centered)$d) / sqrt(nrow(X)) * 2.5
    }
    
    if (length(scale) == 1)
        scale <- rep(scale, ncol(X))
    
    names(scale) <- NULL
    
    assertthat::assert_that( length(scale) == ncol(X) )
    
    
    # Extra axes
    
    if (!is.null(extraAxes)) {
        extraAxes <- as.matrix(extraAxes)
    
        assertthat::assert_that( assertthat::noNA(extraAxes) )
        assertthat::assert_that( nrow(extraAxes) == ncol(X) )
        
        if (is.null(colnames(extraAxes)))
            colnames(extraAxes) <- paste0("E", seq_len(ncol(extraAxes)))
    }
    
    
    # Subsampling
    
    if (!is.null(subsample) && subsample < nrow(X)) {
        ind <- sample.int(nrow(X), subsample)
        
        X <- X[ind,,drop=FALSE]
        
        group <- group[ind]
        
        if (!is.null(name))
            name <- name[ind]
        
        if (!is.null(crosstalkKey))
            crosstalkKey <- crosstalkKey[ind]
        
        if (!is.null(lineFrom)) {
            lineFrom <- match(lineFrom, ind)
            lineTo <- match(lineTo, ind)
            keep <- !is.na(lineFrom) & !is.na(lineTo)
            lineFrom <- lineFrom[keep]
            lineTo <- lineTo[keep]
        }
    }
    
    
    # Convert to form that will JSON-ify correctly
    #
    # htmlwidgets uses jsonlite::toJSON with auto_unbox=TRUE.
    # Numeric vectors of length one become numbers, not arrays.
    # I convert vectors to lists to prevent this.
    
    data <- list(
        X = X,
        center = as.list(as.numeric(center)),
        scale = as.list(as.numeric(scale)),
        colnames = as.list(columnNames),
        rownames = as.list(as.character(name)),
        group = as.list(as.integer(group)-1),
        levels = as.list(levels(group)),
        
        extraAxes = extraAxes,
        extraAxesNames = as.list(colnames(extraAxes)),
        
        # Convert from 1 based to 0 based indices
        lineFrom = as.list(as.numeric(lineFrom) - 1), 
        lineTo = as.list(as.numeric(lineTo) - 1),
        lineColors = as.list(as.character(lineColors)),
        
        axisColors=as.list(as.character(axisColors)),
        levelColors=as.list(as.character(levelColors)),
        colorVariation=as.numeric(colorVariation),
        pointSize=as.numeric(pointSize), # Javascript side can handle unboxing here (number|number[])
        
        crosstalkGroup=crosstalkGroup,
        crosstalkKey=crosstalkKey,
        crosstalkWantFilter=link_filter,
        
        state=state)
    
    
    htmlwidgets::createWidget(
        name = 'langevitour',
        x = data,
        width = width,
        height = height,
        package = 'langevitour',
        elementId = elementId,
        sizingPolicy = htmlwidgets::sizingPolicy(
            defaultWidth = 700,
            defaultHeight = 600,
            viewer.padding = 5,
            browser.fill = TRUE,
            knitr.figure = FALSE
        ),
        dependencies = dependencies
    )
}

#' Shiny bindings for langevitour
#'
#' Output and render functions for using langevitour within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a langevitour, usually a block of code ending with a call to \code{langevitour()}
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name langevitour-shiny
#'
#' @examples
#' 
#' library(shiny)
#' library(palmerpenguins)
#' 
#' completePenguins <- na.omit(penguins[,c(1,3,4,5,6)])
#' scale <- apply(completePenguins[,-1], 2, sd)*4
#' 
#' ui <- fluidPage(
#'     sliderInput('zoom', 'Zoom', 0, min=-1, max=1, step=0.1),
#'     langevitourOutput('widget')
#' )
#' 
#' server <- function(input,output) { 
#'     output$widget <- renderLangevitour({
#'         langevitour(
#'             completePenguins[,-1], 
#'             completePenguins$species, 
#'             scale=scale * 10^input$zoom, pointSize=2)
#'     })
#' }
#' 
#' app <- shinyApp(ui, server)
#' 
#' # Use runApp(app) or runGadget(app) to run app.
#'
#' @export
langevitourOutput <- function(outputId, width = '100%', height = '600px'){
    htmlwidgets::shinyWidgetOutput(outputId, 'langevitour', width, height, package = 'langevitour')
}

#' @rdname langevitour-shiny
#' @export
renderLangevitour <- function(expr, env = parent.frame(), quoted = FALSE) {
    if (!quoted) { expr <- substitute(expr) } # force quoted
    htmlwidgets::shinyRenderWidget(expr, langevitourOutput, env, quoted = TRUE)
}

File: py/langevitour/langevitour.py
import json

import numpy as np
import pkg_resources
from numpy.linalg import svd

from langevitour.utils import as_factor


class Langevitour:
    """
    Langevin Tour class for creating an HTML widget to explore high-dimensional numerical datasets.

    This class provides an interface to visualize high-dimensional datasets using an interactive Langevin Tour widget.
    The main data argument is `data`, while other parameters allow customization of the visualization.

    Args:
        data (np.array): The primary dataset to plot. Rows represent points in the widget and columns represent data variables.
        group (list, optional): A grouping for each row in `data` to color points. Defaults to None.
        column_names (list[str], optional): Names for the data columns. Defaults to None.
        levels (list, optional): Specifies the levels if the group is not None. Defaults to unique values of the group.
        name (str, optional): A name for each row in `data`. Defaults to None.
        center (list, optional): Center for each variable. If omitted, column means are used.
        scale (list, optional): Scale +/- center to determine the range of visible data. Defaults to a standard value.
        extra_axes (np.array, optional): Columns defining projections of interest.
        extra_axes_names (list[str], optional): Names for the extra axes.
        line_from (list[int], optional): Row numbers indicating start of lines.
        line_to (list[int], optional): Row numbers indicating end of lines.
        line_colors (list[str], optional): CSS colors for each line.
        axis_colors (list[str], optional): CSS colors for each variable and extra axis.
        level_colors (list[str], optional): CSS colors for each level of `group`.
        color_variation (float, 0-1, optional): Brightness variation among individual points. Defaults to 0.3.
        point_size (float|list[float], optional): Point radius in pixels. Defaults to 1. A single value or a value for each point.
        subsample (int, optional): Randomly subsample to this many rows for speed. Defaults to None.
        state (str, optional): Initial widget state settings. Defaults to None.
        width (int, optional): Width of widget in pixels. Defaults to 700.
        height (int, optional): Height of widget in pixels. Defaults to 600.
        element_id (str, optional): A unique ID for the widget element. Defaults to None.

    Note:
        In JavaScript, the Langevitour object can be accessed using `document.getElementById(elementId).langevitour`.
        A button could, for example, set the widget state with `document.getElementById(elementId).langevitour.setState(desiredState)`.
    """

    def __init__(
        self,
        data,
        group=None,
        column_names=None,
        levels=None,
        name=None,
        center=None,
        scale=None,
        extra_axes=None,
        extra_axes_names=None,
        line_from=None,
        line_to=None,
        line_colors=None,
        axis_colors=None,
        level_colors=None,
        color_variation=0.1,
        point_size=1,
        subsample=None,
        state=None,
        width=700,
        height=600,
        element_id=None,
    ):
        if hasattr(data, "columns") and column_names is None:
            column_names = list(data.columns)

        # Ensure data is matrix
        data = np.array(data)
        if data.ndim != 2:
            msg = "data must be 2-dimensional."
            raise ValueError(msg)

        # Column names
        if column_names is None:
            column_names = [f"V{i+1}" for i in range(data.shape[1])]

        # Grouping
        if group is None:
            group = ["" for _ in range(data.shape[0])]
        else:
            group = list(map(str, group))

        group, group_levels = as_factor(group, levels)

        if levels is None:
            levels = group_levels

        # Centering
        if center is None:
            center = np.mean(data, axis=0)
        if len(center) == 1:
            center = [center[0] for _ in range(data.shape[1])]

        # Scaling
        if scale is None:
            data_centered = data - center
            scale = max(svd(data_centered, compute_uv=False)) / np.sqrt(data.shape[0]) * 2.5
        if np.isscalar(scale):
            scale = [scale for _ in range(data.shape[1])]

        # Check for problems
        if data.shape[1] < 2:
            msg = "The number of columns in data should be at least 2."
            raise ValueError(msg)

        if len(column_names) != data.shape[1]:
            msg = "Mismatch between number of column names and number of columns in data."
            raise ValueError(msg)

        if len(group) != data.shape[0]:
            msg = "Length of group does not match the number of rows in data."
            raise ValueError(msg)


        # Extra axes
        if extra_axes is not None:
            extra_axes = np.array(extra_axes)
            if extra_axes.shape[0] != data.shape[1]:
                msg = "The number of rows in extra_axes does not match the number of columns in data."
                raise ValueError(msg)
            if extra_axes_names is None:
                extra_axes_names = (
                    list(data.columns) if hasattr(data, "columns") else [f"E{i+1}" for i in range(extra_axes.shape[1])]
                )

        # Subsampling
        if subsample is not None and subsample < data.shape[0]:
            indices = np.random.choice(data.shape[0], subsample, replace=False)
            data = data[indices, :]
            group = [group[i] for i in indices]
            if name:
                name = [name[i] for i in indices]

        self.data = data
        self.group = group
        self.levels = levels
        self.name = name
        self.center = center
        self.scale = scale
        self.column_names = column_names
        self.extra_axes = extra_axes
        self.extra_axes_names = extra_axes_names
        self.line_from = line_from
        self.line_to = line_to
        self.line_colors = line_colors
        self.axis_colors = axis_colors
        self.level_colors = level_colors
        self.color_variation = color_variation
        self.point_size = point_size
        self.subsample = subsample
        self.state = state
        self.width = width
        self.height = height
        self.element_id = element_id

    def _get_js_content(self):
        """
        Retrieve the JavaScript content for the Langevintour widget.
        Returns:
            str: The JavaScript content for the widget.
        """
        js_path = pkg_resources.resource_filename("langevitour", "static/langevitour-pack.js")
        with open(js_path) as file:
            return file.read()

    def to_json(self):
        """
        Convert the Langevintour object to a JSON string.
        Returns:
            str: A JSON string representation of the object.
        """
        # Create a dictionary representation of the instance variables
        data_dict = {
            "X": self.data.tolist()
            if isinstance(self.data, (np.ndarray, np.generic))
            else self.data,  # Convert numpy arrays to list
            "group": self.group,
            "levels": self.levels,
            "name": self.name,
            "center": self.center.tolist()
            if isinstance(self.center, (np.ndarray, np.generic))
            else self.center,  # Convert numpy arrays to list
            "scale": self.scale,
            "colnames": self.column_names,
            "extraAxes": self.extra_axes.tolist()
            if isinstance(self.extra_axes, (np.ndarray, np.generic))
            else self.extra_axes,  # Convert numpy arrays to list
            "extraAxesNames": self.extra_axes_names,
            "lineFrom": self.line_from,
            "lineTo": self.line_to,
            "lineColors": self.line_colors,
            "axisColors": self.axis_colors,
            "levelColors": self.level_colors,
            "colorVariation": self.color_variation,
            "pointSize": self.point_size,
            "subsample": self.subsample,
            "state": self.state,
            "width": self.width,
            "height": self.height,
            "elementId": self.element_id,
        }

        # Convert the dictionary to a JSON string and return
        return json.dumps(data_dict, default=str)

    def to_html(self):
        """
        Convert the Langevin Tour object to an HTML string.
        Returns:
            str: An HTML string representation of the object.
        """
        import time
        unique_id = f"langevi_tour_{int(time.time() * 1000)}" if self.element_id is None else self.element_id
        data_json = self.to_json()
        js_content = self._get_js_content()

        html = f"""
        <div id="{unique_id}"></div>
        <script>
            if ('el_{unique_id}' in window) {{
                // ensure that the previous instance is cleaned up
                delete window.el_{unique_id};
                delete window.tour_{unique_id};
            }}

            {js_content}

            var el_{unique_id} = document.getElementById("{unique_id}");
            var tour_{unique_id} = new langevitour.Langevitour(el_{unique_id}, {self.width}, {self.height});
            tour_{unique_id}.renderValue({data_json});
        </script>
        """
        return html


    def _repr_html_(self):
        """
        Represent the Langevintour object as an HTML string.
        Used by jupyter notebooks to render the tour inline.
        Returns:
            str: An HTML string representation of the object.
        """
        return self.to_html()

    def write_html(self, filename="langevitour_plot.html"):
        """
        Write the Langevintour object to an HTML file.
        Args:
            filename (str, optional): The name of the HTML file. Defaults to 'langevitour_plot.html'.
        Side Effects:
            Writes the HTML representation of the object to the specified file.
        """
        with open(filename, "w") as file:
            file.write(self._repr_html_())
